File: src/main/java/org/fountainmc/api/entity/EntityType.java
package org.fountainmc.api.entity;

import org.fountainmc.api.Fountain;
import org.fountainmc.api.entity.hanging.ItemFrame;
import org.fountainmc.api.entity.hanging.LeashKnot;
import org.fountainmc.api.entity.hanging.Painting;

public interface EntityType<T extends Entity> {

    EntityType<Player> PLAYER = Fountain.getServer().getEntityType("Player", Player.class);
    EntityType<Entity> XP_ORB = Fountain.getServer().getEntityType("XPOrb", Entity.class);
    EntityType<Entity> AREA_EFFECT_CLOUD = Fountain.getServer().getEntityType("AreaEffectCloud", Entity.class);
    EntityType<Entity> ENDER_SIGNAL = Fountain.getServer().getEntityType("EyeOfEnderSignal", Entity.class);
    EntityType<Entity> PRIMED_TNT = Fountain.getServer().getEntityType("PrimedTnt", Entity.class);
    EntityType<Entity> FALLING_BLOCK = Fountain.getServer().getEntityType("FallingSand", Entity.class);
    EntityType<Entity> ARMOR_STAND = Fountain.getServer().getEntityType("ArmorStand", Entity.class);
    EntityType<Entity> ENDER_CRYSTAL = Fountain.getServer().getEntityType("EnderCrystal", Entity.class);

    // Hanging entities
    EntityType<LeashKnot> LEASH_KNOT = Fountain.getServer().getEntityType("LeashKnot", LeashKnot.class);
    EntityType<Painting> PAINTING = Fountain.getServer().getEntityType("Painting", Painting.class);
    EntityType<ItemFrame> ITEM_FRAME = Fountain.getServer().getEntityType("ItemFrame", ItemFrame.class);

    // Projectiles
    EntityType<Entity> FIREWORK_ROCKET = Fountain.getServer().getEntityType("FireworksRocketEntity", Entity.class);
    EntityType<Entity> SPECTRAL_ARROW = Fountain.getServer().getEntityType("SpectralArrow", Entity.class);
    EntityType<Entity> SHULKER_BULLET = Fountain.getServer().getEntityType("ShulkerBullet", Entity.class);
    EntityType<Entity> DRAGON_FIREBALL = Fountain.getServer().getEntityType("DragonFireball", Entity.class);
    EntityType<Entity> WITHER_SKULL = Fountain.getServer().getEntityType("WitherSkull", Entity.class);
    EntityType<Entity> THROWN_POTION = Fountain.getServer().getEntityType("ThrownPotion", Entity.class);
    EntityType<Entity> THROWN_EXP_BOTTLE = Fountain.getServer().getEntityType("ThrownExpBottle", Entity.class);
    EntityType<Entity> ARROW = Fountain.getServer().getEntityType("Arrow", Entity.class);
    EntityType<Entity> SNOWBALL = Fountain.getServer().getEntityType("Snowball", Entity.class);
    EntityType<Entity> FIREBALL = Fountain.getServer().getEntityType("Fireball", Entity.class);
    EntityType<Entity> SMALL_FIREBALL = Fountain.getServer().getEntityType("SmallFireball", Entity.class);
    EntityType<Entity> THROWN_ENDERPEARL = Fountain.getServer().getEntityType("ThrownEnderpearl", Entity.class);
    EntityType<Entity> THROWN_EGG = Fountain.getServer().getEntityType("ThrownEgg", Entity.class);

    // Vehicles
    EntityType<Entity> BOAT = Fountain.getServer().getEntityType("Boat", Entity.class);

    // TODO: MINECARTS
    //
    // Living Entities
    //
    EntityType<LivingEntity> CREEPER = Fountain.getServer().getEntityType("Creeper", LivingEntity.class);
    EntityType<LivingEntity> SKELETON = Fountain.getServer().getEntityType("Skeleton", LivingEntity.class);
    EntityType<LivingEntity> SPIDER = Fountain.getServer().getEntityType("Spider", LivingEntity.class);
    EntityType<LivingEntity> GIANT = Fountain.getServer().getEntityType("Giant", LivingEntity.class);
    EntityType<LivingEntity> ZOMBIE = Fountain.getServer().getEntityType("Zombie", LivingEntity.class);
    EntityType<LivingEntity> SLIME = Fountain.getServer().getEntityType("Slime", LivingEntity.class);
    EntityType<LivingEntity> GHAST = Fountain.getServer().getEntityType("Ghast", LivingEntity.class);
    EntityType<LivingEntity> PIG_ZOMBIE = Fountain.getServer().getEntityType("PigZombie", LivingEntity.class);
    EntityType<LivingEntity> ENDERMAN = Fountain.getServer().getEntityType("Enderman", LivingEntity.class);
    EntityType<LivingEntity> CAVE_SPIDER = Fountain.getServer().getEntityType("CaveSpider", LivingEntity.class);
    EntityType<LivingEntity> SILVERFISH = Fountain.getServer().getEntityType("Silverfish", LivingEntity.class);
    EntityType<LivingEntity> BLAZE = Fountain.getServer().getEntityType("Blaze", LivingEntity.class);
    EntityType<LivingEntity> MAGMA_CUBE = Fountain.getServer().getEntityType("LavaSlime", LivingEntity.class);
    EntityType<LivingEntity> ENDER_DRAGON = Fountain.getServer().getEntityType("EnderDragon", LivingEntity.class);
    EntityType<LivingEntity> WITHER = Fountain.getServer().getEntityType("WitherBoss", LivingEntity.class);
    EntityType<LivingEntity> BAT = Fountain.getServer().getEntityType("Bat", LivingEntity.class);
    EntityType<LivingEntity> WITCH = Fountain.getServer().getEntityType("Witch", LivingEntity.class);
    EntityType<LivingEntity> ENDERMITE = Fountain.getServer().getEntityType("Endermite", LivingEntity.class);
    EntityType<LivingEntity> GUARDIAN = Fountain.getServer().getEntityType("Guardian", LivingEntity.class);
    EntityType<LivingEntity> SHULKER = Fountain.getServer().getEntityType("Shulker", LivingEntity.class);
    EntityType<LivingEntity> PIG = Fountain.getServer().getEntityType("Pig", LivingEntity.class);
    EntityType<LivingEntity> SHEEP = Fountain.getServer().getEntityType("Sheep", LivingEntity.class);
    EntityType<LivingEntity> COW = Fountain.getServer().getEntityType("Cow", LivingEntity.class);
    EntityType<LivingEntity> CHICKEN = Fountain.getServer().getEntityType("Chicken", LivingEntity.class);
    EntityType<LivingEntity> SQUID = Fountain.getServer().getEntityType("Squid", LivingEntity.class);
    EntityType<LivingEntity> WOLF = Fountain.getServer().getEntityType("Wolf", LivingEntity.class);
    EntityType<LivingEntity> MUSHROOM_COW = Fountain.getServer().getEntityType("MushroomCow", LivingEntity.class);
    EntityType<LivingEntity> SNOWMAN = Fountain.getServer().getEntityType("SnowMan", LivingEntity.class);
    EntityType<LivingEntity> OCELOT = Fountain.getServer().getEntityType("Ozelot", LivingEntity.class);
    EntityType<LivingEntity> IRON_GOLEM = Fountain.getServer().getEntityType("VillagerGolem", LivingEntity.class);
    EntityType<LivingEntity> HORSE = Fountain.getServer().getEntityType("EntityHorse", LivingEntity.class);
    EntityType<LivingEntity> RABBIT = Fountain.getServer().getEntityType("Rabbit", LivingEntity.class);
    EntityType<LivingEntity> VILLAGER = Fountain.getServer().getEntityType("Villager", LivingEntity.class);

    /**
     * Get the maximum health a entity should have
     *
     * @return the maximum health
     */
    double getMaxHealth();

    /**
     * Get the class that implements this entity type
     *
     * @return the class that implements this entity type
     */
    Class<? extends T> getEntityClass();

}

File: src/main/kotlin/org/fountainmc/api/entity/EntityType.kt
package org.fountainmc.api.entity

import org.fountainmc.api.Fountain
import org.fountainmc.api.entity.hanging.ItemFrame
import org.fountainmc.api.entity.hanging.LeashKnot
import org.fountainmc.api.entity.hanging.Painting
import kotlin.reflect.KClass
import kotlin.reflect.full.isSuperclassOf

interface EntityType<out T : Entity> {

    /**
     * The maximum health this type of entity should have.
     */
    val maxHealth: Double

    /**
     * The class that implements this entity type
     */
    val entityClass: KClass<out T>

    companion object {

        val PLAYER = Fountain.getEntityTypeAs("Player", Player::class)
        val XP_ORB = Fountain.getEntityTypeAs("XPOrb", Entity::class)
        val AREA_EFFECT_CLOUD = Fountain.getEntityTypeAs("AreaEffectCloud", Entity::class)
        val ENDER_SIGNAL = Fountain.getEntityTypeAs("EyeOfEnderSignal", Entity::class)
        val PRIMED_TNT = Fountain.getEntityTypeAs("PrimedTnt", Entity::class)
        val FALLING_BLOCK = Fountain.getEntityTypeAs("FallingSand", Entity::class)
        val ARMOR_STAND = Fountain.getEntityTypeAs("ArmorStand", Entity::class)
        val ENDER_CRYSTAL = Fountain.getEntityTypeAs("EnderCrystal", Entity::class)

        // Hanging entities
        val LEASH_KNOT = Fountain.getEntityTypeAs("LeashKnot", LeashKnot::class)
        val PAINTING = Fountain.getEntityTypeAs("Painting", Painting::class)
        val ITEM_FRAME = Fountain.getEntityTypeAs("ItemFrame", ItemFrame::class)

        // Projectiles
        val FIREWORK_ROCKET = Fountain.getEntityTypeAs("FireworksRocketEntity", Entity::class)
        val SPECTRAL_ARROW = Fountain.getEntityTypeAs("SpectralArrow", Entity::class)
        val SHULKER_BULLET = Fountain.getEntityTypeAs("ShulkerBullet", Entity::class)
        val DRAGON_FIREBALL = Fountain.getEntityTypeAs("DragonFireball", Entity::class)
        val WITHER_SKULL = Fountain.getEntityTypeAs("WitherSkull", Entity::class)
        val THROWN_POTION = Fountain.getEntityTypeAs("ThrownPotion", Entity::class)
        val THROWN_EXP_BOTTLE = Fountain.getEntityTypeAs("ThrownExpBottle", Entity::class)
        val ARROW = Fountain.getEntityTypeAs("Arrow", Entity::class)
        val SNOWBALL = Fountain.getEntityTypeAs("Snowball", Entity::class)
        val FIREBALL = Fountain.getEntityTypeAs("Fireball", Entity::class)
        val SMALL_FIREBALL = Fountain.getEntityTypeAs("SmallFireball", Entity::class)
        val THROWN_ENDERPEARL = Fountain.getEntityTypeAs("ThrownEnderpearl", Entity::class)
        val THROWN_EGG = Fountain.getEntityTypeAs("ThrownEgg", Entity::class)

        // Vehicles
        val BOAT = Fountain.getEntityTypeAs("Boat", Entity::class)

        // TODO: MINECARTS
        //
        // Living Entities
        //
        val CREEPER = Fountain.getEntityTypeAs("Creeper", LivingEntity::class)
        val SKELETON = Fountain.getEntityTypeAs("Skeleton", LivingEntity::class)
        val SPIDER = Fountain.getEntityTypeAs("Spider", LivingEntity::class)
        val GIANT = Fountain.getEntityTypeAs("Giant", LivingEntity::class)
        val ZOMBIE = Fountain.getEntityTypeAs("Zombie", LivingEntity::class)
        val SLIME = Fountain.getEntityTypeAs("Slime", LivingEntity::class)
        val GHAST = Fountain.getEntityTypeAs("Ghast", LivingEntity::class)
        val PIG_ZOMBIE = Fountain.getEntityTypeAs("PigZombie", LivingEntity::class)
        val ENDERMAN = Fountain.getEntityTypeAs("Enderman", LivingEntity::class)
        val CAVE_SPIDER = Fountain.getEntityTypeAs("CaveSpider", LivingEntity::class)
        val SILVERFISH = Fountain.getEntityTypeAs("Silverfish", LivingEntity::class)
        val BLAZE = Fountain.getEntityTypeAs("Blaze", LivingEntity::class)
        val MAGMA_CUBE = Fountain.getEntityTypeAs("LavaSlime", LivingEntity::class)
        val ENDER_DRAGON = Fountain.getEntityTypeAs("EnderDragon", LivingEntity::class)
        val WITHER = Fountain.getEntityTypeAs("WitherBoss", LivingEntity::class)
        val BAT = Fountain.getEntityTypeAs("Bat", LivingEntity::class)
        val WITCH = Fountain.getEntityTypeAs("Witch", LivingEntity::class)
        val ENDERMITE = Fountain.getEntityTypeAs("Endermite", LivingEntity::class)
        val GUARDIAN = Fountain.getEntityTypeAs("Guardian", LivingEntity::class)
        val SHULKER = Fountain.getEntityTypeAs("Shulker", LivingEntity::class)
        val PIG = Fountain.getEntityTypeAs("Pig", LivingEntity::class)
        val SHEEP = Fountain.getEntityTypeAs("Sheep", LivingEntity::class)
        val COW = Fountain.getEntityTypeAs("Cow", LivingEntity::class)
        val CHICKEN = Fountain.getEntityTypeAs("Chicken", LivingEntity::class)
        val SQUID = Fountain.getEntityTypeAs("Squid", LivingEntity::class)
        val WOLF = Fountain.getEntityTypeAs("Wolf", LivingEntity::class)
        val MUSHROOM_COW = Fountain.getEntityTypeAs("MushroomCow", LivingEntity::class)
        val SNOWMAN = Fountain.getEntityTypeAs("SnowMan", LivingEntity::class)
        val OCELOT = Fountain.getEntityTypeAs("Ozelot", LivingEntity::class)
        val IRON_GOLEM = Fountain.getEntityTypeAs("VillagerGolem", LivingEntity::class)
        val HORSE = Fountain.getEntityTypeAs("EntityHorse", LivingEntity::class)
        val RABBIT = Fountain.getEntityTypeAs("Rabbit", LivingEntity::class)
        val VILLAGER = Fountain.getEntityTypeAs("Villager", LivingEntity::class)
    }

}

fun <T: Entity> EntityType<*>.cast(entityClass: KClass<T>): EntityType<T> {
    require(entityClass.isSuperclassOf(this.entityClass)) { "Entity type $this is not a $entityClass" }
    @Suppress("UNCHECKED_CAST") // We checked ;)
    return this as EntityType<T>
}
